trax/models/research/funnel_transformer.py

```diff
@@ -0,0 +1,269 @@
+
+
+def _Upsample(short, masks, long):
+  factor = -(-long.shape[1] // short.shape[1])  # ceil division
+  new_vecs = long + short.repeat(factor, axis=1)[:, :long.shape[1], :]
+  new_masks = masks.repeat(factor, axis=-1)[:, :, :, :long.shape[1]]
+  return new_vecs, new_masks
+
+
+def _Upsampler():
+  return tl.Fn('Upsampler', _Upsample, n_out=2)
+
+
+def _FunnelBlock(d_model, d_ff, n_heads,
+                 dropout, dropout_shared_axes, mode, ff_activation,
+                 pool_layer, pool_size, strides, separate_cls):
+  """Internal funnel block. On input it takes (activations, masks).
+
+  Args:
+      d_model: Final dimension of tensors at most points in the model, including
+          the initial embedding output.
+      d_ff: Size of special dense layer in the feed-forward part of each block.
+      n_heads: Number of attention heads.
+      dropout: Stochastic rate (probability) for dropping an activation value
+          when applying dropout within a block.
+      dropout_shared_axes: Tensor axes on which to share a dropout mask.
+          Sharing along batch and sequence axes (`dropout_shared_axes=(0,1)`) is
+          a useful way to save memory and apply consistent masks to activation
+          vectors at different sequence positions.
+      mode: If `'train'`, each block will include dropout; else, it will
+          pass all values through unaltered.
+      ff_activation: Type of activation function at the end of each block; must
+          be an activation-type subclass of `Layer`.
+      pool_size: Shape of window that gets reduced to a single vector value.
+          If the layer inputs are :math:`n`-dimensional arrays, then `pool_size`
+          must be a tuple of length :math:`n-2`.
+      strides: Offsets from the location of one window to the locations of
+          neighboring windows along each axis. If specified, must be a tuple of
+          the same length as `pool_size`. If None, then offsets of 1 along each
+          window axis, :math:`(1, ..., 1)`, will be used.
+  Returns:
+      A list of layers that maps (activations, mask) to (activations', mask).
+  """
+  attention = tl.AttentionQKV(
+      d_feature=d_model, n_heads=n_heads, dropout=dropout, mode=mode)
+  feed_forward = _FeedForwardBlock(
+      d_model, d_ff, dropout, dropout_shared_axes, mode, ff_activation)
+  pooling = PoolLayer(pool_layer, pool_size, strides, separate_cls)
+
+  return tl.Serial(                     # h, mask
+      tl.Branch(pooling, None, None),   # h', h, h, mask
+      tl.Dup(),                         # h', h', h, h, mask
+      tl.Parallel(
+          None,
+          attention
+      ),                                # h', attention(...), mask
+      tl.Add(),                         # h'+attention(...), mask
+      tl.LayerNorm(),                   # funnel_activations, mask
+      tl.Parallel(
+          None,
+          tl.Fn('mask_max_pool',
+                _InternalMaxPool),
```

Comment:
This pool layer used for masks doesn't have pool_size/strides synchronized with pool layer for attention. So, setting pool_size/strides different than 2 will not work as I understand it.
Can you this pool layer to change mask according to pool_size?
(This comment applies also to _FunnelResidualBlock. )